File: munchly/lib/features/bookings/domain/entities/booking.dart
import 'package:equatable/equatable.dart';

/// Booking entity
class Booking extends Equatable {
  final String id;
  final String restaurantId;
  final String restaurantName;
  final String tableId;
  final int tableNumber;
  final String userId;
  final DateTime date;
  final String time;
  final int guests;
  final BookingStatus status;

  const Booking({
    required this.id,
    required this.restaurantId,
    required this.restaurantName,
    required this.tableId,
    required this.tableNumber,
    required this.userId,
    required this.date,
    required this.time,
    required this.guests,
    required this.status,
  });

  @override
  List<Object?> get props => [
        id,
        restaurantId,
        restaurantName,
        tableId,
        tableNumber,
        userId,
        date,
        time,
        guests,
        status,
      ];

  /// Copy with method for updating booking
  Booking copyWith({
    String? id,
    String? restaurantId,
    String? restaurantName,
    String? tableId,
    int? tableNumber,
    String? userId,
    DateTime? date,
    String? time,
    int? guests,
    BookingStatus? status,
  }) {
    return Booking(
      id: id ?? this.id,
      restaurantId: restaurantId ?? this.restaurantId,
      restaurantName: restaurantName ?? this.restaurantName,
      tableId: tableId ?? this.tableId,
      tableNumber: tableNumber ?? this.tableNumber,
      userId: userId ?? this.userId,
      date: date ?? this.date,
      time: time ?? this.time,
      guests: guests ?? this.guests,
      status: status ?? this.status,
    );
  }
}

/// Booking status enum
enum BookingStatus {
  pending,
  confirmed,
  cancelled,
  completed;

  String get displayName {
    switch (this) {
      case BookingStatus.pending:
        return 'Pending';
      case BookingStatus.confirmed:
        return 'Confirmed';
      case BookingStatus.cancelled:
        return 'Cancelled';
      case BookingStatus.completed:
        return 'Completed';
    }
  }
}

File: backend/src/models/Booking.js
const mongoose = require('mongoose');

const bookingSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'User ID is required'],
    },
    restaurantId: {
      type: String,
      required: [true, 'Restaurant ID is required'],
    },
    restaurantName: {
      type: String,
      required: [true, 'Restaurant name is required'],
    },
    tableId: {
      type: String,
      required: [true, 'Table ID is required'],
    },
    tableNumber: {
      type: Number,
      required: [true, 'Table number is required'],
    },
    date: {
      type: String,
      required: [true, 'Date is required'],
      match: [/^\d{4}-\d{2}-\d{2}$/, 'Date must be in YYYY-MM-DD format'],
    },
    time: {
      type: String,
      required: [true, 'Time is required'],
      match: [/^\d{2}:\d{2}$/, 'Time must be in HH:MM format'],
    },
    guests: {
      type: Number,
      required: [true, 'Number of guests is required'],
      min: [1, 'At least 1 guest is required'],
    },
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'cancelled', 'completed'],
      default: 'confirmed',
    },
  },
  {
    timestamps: true,
  }
);

// Transform output to match API spec
bookingSchema.methods.toJSON = function () {
  const obj = this.toObject();
  obj.id = obj._id.toString();
  obj.userId = obj.userId.toString();
  delete obj._id;
  delete obj.__v;
  delete obj.updatedAt;
  return obj;
};

const Booking = mongoose.model('Booking', bookingSchema);

module.exports = Booking;
